tailassignment_oop.py

```diff
@@ -0,0 +1,376 @@
+from qaoa_OOP import *
+
+class QAOATailAssignment(QAOAStandard):
+
+    def __init__(self,options):
+        super().__init__(options['CR'].size, options)
+
+        # Vectorized functions for the cost and is_solution:
+        # Very useful for simulation with the statevector, as it
+        # allows for avoiding multiple nested loops in the measurementStatistics function
+        
+        self.vector_cost = np.vectorize(self.cost)
+        self.vector_is_solution = np.vectorize(self.is_solution)
+
+        # Save matrices as variables of object:
+        
+        self.FR         = self.options.get('FR', None)
+        self.CR         = self.options.get('CR', None)
+        self.mu         = self.options.get('mu', 1)
+
+        self.F, self.R  = np.shape(self.FR)
+
+
+    def cost(self,binstring):
+        
+        # Reverse string since qiskit uses ordering MSB ... LSB
+        x = np.array(list(map(int,binstring[::-1])))
+        return - ( (self.CR @ x) + self.mu * np.sum((1 - (self.FR @ x))**2) )
+
+    def mix_states(self,beta):
+        """
+        Applies unitary evolution of mixer hamiltonian with 
+        time parameter beta
+
+        Parameters
+        ----------
+        beta : float
+            Time/angle for applying hamiltonian.
+
+        """
+        self.qc.rx( - 2 * beta, self.q_register ) 
+
+    def apply_exco(self,gamma):
+        """
+        Applies unitary evolution of exact cover hamiltonian with 
+        time-parameter gamma
+
+        Parameters
+        ----------
+        gamma : float
+            Time/angle for applying hamiltonian.
+
+        """
+        
+        for r in range(self.R):
+            hr = 0.5 * self.mu * self.FR[:,r] @ (np.sum(self.FR,axis = 1) - 2)
+
+            self.qc.rz( gamma * hr, self.q_register[r])
+
+            for r_ in range(r+1,self.R):
+                Jrr_  = 0.5 * self.FR[:,r] @ self.FR[:,r_]
+
+                # Apply U(gamma), coupling part
+
+                self.qc.cx(self.q_register[r], self.q_register[r_])
+                self.qc.rz(gamma * Jrr_, self.q_register[r_])
+                self.qc.cx(self.q_register[r], self.q_register[r_])
+
+    def apply_cost(self,gamma):
+        """
+        Applies unitary evolution of cost hamiltonian with 
+        time-parameter gamma
+
+        Parameters
+        ----------
+        gamma : float
+            Time/angle for applying hamiltonian.
+
+        """
+
+        for r in range(self.R):
+            hr = 0.5 * self.CR[r]
+
+            self.qc.rz(gamma * hr, self.q_register[r])
+
+    def apply_hamiltonian(self,gamma):
```

Comment:
is this necessary, can't his reuse the apply functions from above?

Reply:
Strictly speaking, no it is not required. But applying the two function above would require more gates, which I commented on in the `createCircuit` function of the tailassignment class. Having this function only makes it possible to create a more compact circuit.